IPNS.md

```diff
@@ -29,95 +36,266 @@ Each time a file is modified, its content address changes. As a consequence, the
+
+The means of distribution are left unspecified. Implementations MAY choose to
+publish signed record using multiple routing systems, such as
+[libp2p Kademlia DHT](https://github.com/libp2p/specs/tree/master/kad-dht),
+[PubSub topic](naming/pubsub.md), or a [Reframe endpoint](reframe/) (see [Routing record](#routing-record)).
 
-## API Spec
+On the other side, each peer must be able to get a record published by another node. It only needs to have the unique identifier used to publish the record to the network. Taking into account the routing system being used, we may obtain a set of occurrences of the record from the network. In this case, records can be compared using the sequence number, in order to obtain the most recent one.
 
-  - [API_CORE](https://github.com/ipfs/specs/blob/master/API_CORE.md)
+As soon as the node has the most recent record, the signature and the validity must be [verified](#record-verification), in order to conclude that the record is still valid and not compromised.
 
-## Implementations
+Finally, the network nodes may also republish their records, so that the records in the network continue to be valid to the other nodes.
 
-  - [js-ipfs](https://github.com/ipfs/js-ipfs/tree/master/packages/ipfs-core/src/ipns)
-  - [go-namesys](https://github.com/ipfs/go-namesys)
+### Record Creation
+
+IPNS record MUST be serialized as `IpnsEntry` protobuf and `IpfsEntry.data` MUST be signed using the private key.
+
+Creating a new IPNS record MUST follow the below steps:
+
+1. Create `IpnsEntry` and set `value`, `validity`, `validityType`, `sequence`, and `ttl`
+   - If you are updating an existing record, remember to increase values in `sequence` and `validity`
+2. Create a DAG-CBOR document with the same values for `value`, `validity`, `validityType`, `sequence`, and `ttl`
+   - This is paramount: this CBOR will be used for signing.
+3. Store DAG-CBOR in `IpnsEntry.data`.
+   - If you want to store additional metadata in the record, add it under unique keys at `IpnsEntry.data`.
+   - The order of fields impacts signature verification. If you are using an alternative CBOR implementation, make sure the CBOR field order follows [RFC7049](https://www.rfc-editor.org/rfc/rfc7049) sorting rules: length and then bytewise. The order of fields impacts signature verification.
+4. If your public key can't be inlined inside the IPNS Name, include a serialized copy in `IpnsEntry.pubKey`
+   - This step SHOULD be skipped for Ed25519, and any other key types that are inlined inside of [IPNS Name](#ipns-name) itself.
+6. Create bytes for signing by concatenating `ipns-signature:` prefix (bytes in hex: `69706e732d7369676e61747572653a`) with raw CBOR bytes from `IpnsEntry.data`
+7. Sign concatenated bytes from the previous step using the private key, and store the signature in `IpnsEntry.signatureV2`
+8. Confirm that bytes with serialized `IpnsEntry` are less than or equal 10 kiB in size
```

Comment:
Maybe "Confirm that the serialized `IpnsEntry` bytes sum to less than or equal to 10KiB in size"?